morpho-checkpoint-node/src/Morpho/Node/Env.hs

```diff
@@ -0,0 +1,123 @@
+module Morpho.Node.Env where
+
+import Cardano.BM.Data.Transformers
+import Cardano.BM.Trace
+import Cardano.Crypto.DSIGN
+import Cardano.Prelude
+import Cardano.Shell.Types
+import Control.Monad.Fail
+import qualified Data.Text as T
+import Data.Time
+import Morpho.Config.Logging
+import Morpho.Config.Topology
+import Morpho.Config.Types
+import Morpho.Crypto.ECDSASignature
+import Morpho.Ledger.Block
+import Morpho.Ledger.Update
+import Morpho.Tracing.Tracers
+import Network.HostName
+import Ouroboros.Consensus.BlockchainTime
+import Ouroboros.Consensus.Config
+import Ouroboros.Consensus.Node.ProtocolInfo
+import Ouroboros.Consensus.NodeId
+import Ouroboros.Consensus.Protocol.BFT
+import Ouroboros.Network.Magic
+import Ouroboros.Network.NodeToClient
+import Ouroboros.Network.NodeToNode
+import System.Directory
+import Prelude (error, id)
+
+getHostname :: IO Text
+getHostname = do
+  hn0 <- T.pack <$> getHostName
+  return $ T.take 8 $ fst $ T.breakOn "." hn0
+
+-- | Turns the user configuration (which can be provided by the CLI, config
+-- files or from defaults) into an application environment. This function takes
+-- care of initializing/checking values provided by the config
+configurationToEnv ::
+  ( MorphoStateDefaultConstraints h c,
+    Signable (BftDSIGN c) (MorphoStdHeader h c)
+  ) =>
+  FilePath ->
+  NodeConfiguration ->
+  IO (Env h c, [CardanoFeature])
+configurationToEnv configFile nc = do
+  start <- maybe (SystemStart <$> getCurrentTime) pure (ncSystemStart nc)
+
+  mprivKey <- liftIO . readPrivateKey $ ncNodePrivKeyFile nc
+  privKey <- case mprivKey of
+    Left err -> fail $ "Failed to import private key from " <> show (ncNodePrivKeyFile nc) <> ": " <> show err
+    Right pk -> return pk
+
+  host <- getHostname
+
+  (loggingLayer, loggingFeats) <- loggingFeatures configFile (ncLoggingSwitch nc)
+
+  let basicTrace =
+        setHostname host $
+          appendName "node" (llBasicTrace loggingLayer)
+
+  tracers <- mkTracers (ncTraceOpts nc) basicTrace
+
+  topology <-
+    either error id <$> readTopologyFile (unTopology $ ncTopologyFile nc)
+
+  databaseDir <- canonicalizePath =<< makeAbsolute (unDB $ ncDatabaseDir nc)
+
+  return
+    ( Env
```

Comment:
I do like `RecordWildcards` for destructuring in arguments, though not as much for construction. In this case I think it would make it more complex, since all the `nc*` arguments are different from the `e*` ones.

Reply:
👍 

There is also a destructuring record fields into different bindings but that looks a bit weird...